File: que/src/lossless/consumer.rs
use std::ops::{Deref, DerefMut};
use std::sync::atomic::Ordering;
use std::{ptr::NonNull, sync::Arc};

use bytemuck::AnyBitPattern;
use derivative::Derivative;

use crate::{
    error::QueError, page_size::PageSize, shmem::Shmem, ChannelMode,
    ShmemMode, MAGIC,
};

use super::{burst_amount, Channel};

unsafe impl<M: ChannelMode<T>, T, const N: usize> Send
    for Consumer<M, T, N>
{
}

#[repr(C)]
pub struct Consumer<M: ChannelMode<T>, T, const N: usize> {
    spsc: NonNull<Channel<M, T, N>>,
    head: usize,
    items_since_last_sync: usize,
    consumer_index: usize,
    last_producer_heartbeat: usize,
}

impl<T: AnyBitPattern, const N: usize> Consumer<ShmemMode, T, N> {
    /// Joins an existing channel back by shared memory as a consumer.
    pub unsafe fn join_shmem(
        shmem_id: &str,
        #[cfg(target_os = "linux")] page_size: PageSize,
    ) -> Result<Consumer<ShmemMode, T, N>, QueError> {
        #[cfg(not(target_os = "linux"))]
        let page_size = PageSize::Standard;

        // Calculate buffer size.
        // If using huge pages, we must uplign to page size.
        let buffer_size: i64 = page_size
            .mem_size(core::mem::size_of::<Channel<ShmemMode, T, N>>())
            .try_into()
            .map_err(|_| QueError::InvalidSize)?;

        // Open shmem
        let shmem = Shmem::open_or_create(
            shmem_id,
            buffer_size,
            #[cfg(target_os = "linux")]
            page_size,
        )?;

        unsafe { Consumer::join(shmem.get_mut_ptr()) }
    }
}

#[derive(Derivative)]
#[derivative(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Element<'a, M: ChannelMode<T>, T, const N: usize> {
    value: &'a mut T,
    #[derivative(Debug = "ignore")]
    #[derivative(PartialEq = "ignore")]
    #[derivative(PartialOrd = "ignore")]
    #[derivative(Ord = "ignore")]
    consumer: &'a mut Consumer<M, T, N>,
}

impl<'a, M: ChannelMode<T>, T: PartialEq, const N: usize> PartialEq<T>
    for Element<'a, M, T, N>
{
    fn eq(&self, other: &T) -> bool {
        self.deref().eq(other)
    }
}
impl<'a, M: ChannelMode<T>, T: PartialOrd, const N: usize> PartialOrd<T>
    for Element<'a, M, T, N>
{
    fn partial_cmp(&self, other: &T) -> Option<std::cmp::Ordering> {
        self.deref().partial_cmp(other)
    }
}

impl<'a, M: ChannelMode<T>, T, const N: usize> Drop
    for Element<'a, M, T, N>
{
    fn drop(&mut self) {
        (*self.consumer).head += 1;
        (*self.consumer).items_since_last_sync += 1;
        (*self.consumer).maybe_sync();
    }
}

impl<'a, M: ChannelMode<T>, T, const N: usize> Deref
    for Element<'a, M, T, N>
{
    type Target = T;
    fn deref(&self) -> &Self::Target {
        self.value
    }
}

impl<'a, M: ChannelMode<T>, T, const N: usize> DerefMut
    for Element<'a, M, T, N>
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.value
    }
}

impl<M: ChannelMode<T>, T, const N: usize> Consumer<M, T, N> {
    const MODULO_MASK: usize = N - 1;
    /// Joins an existing channel backed by `buffer`.
    ///
    ///
    /// SAFETY:
    /// This must point to a buffer of proper size and alignment.
    ///
    /// In LocalMode, must point to a region allocated by an Arc with the strong count already incremented!
    pub unsafe fn join(
        buffer: *mut u8,
    ) -> Result<Consumer<M, T, N>, QueError> {
        let buffer = buffer;
        assert!(
            N > 0 && N.is_power_of_two(),
            "Capacity must be a power of two"
        );
        assert!(buffer as usize % 128 == 0, "unaligned");

        // Zerocopy deserialize the SPSC
        let spsc: *mut Channel<M, T, N> = buffer.cast();

        // Check magic
        let magic = (*spsc).magic.load(Ordering::Acquire);
        let capacity = (*spsc).capacity.load(Ordering::Acquire);
        if magic == MAGIC {
            // Check capacity
            if capacity != N {
                return Err(QueError::IncorrectCapacity(capacity));
            }

            // Assume spsc is empty upon joining
            let new_head = (*spsc).tail.load(Ordering::Acquire);
            (*spsc)
                .head
                .store(new_head, Ordering::Release);

            if M::BACKED_BY_ARCC {
                unsafe {
                    Arc::increment_strong_count(spsc);
                }
            }

            // Successful join if magic and capacity is correct
            Ok(Consumer {
                spsc: NonNull::new_unchecked(buffer.cast()),
                head: new_head,
                items_since_last_sync: 0,
                consumer_index: 0,
                last_producer_heartbeat: (*spsc)
                    .producer_heartbeat
                    .load(Ordering::Acquire),
            })
        } else if magic == 0 {
            // Technically could be corrupted but uninitialized
            // is most likely explanation
            Err(QueError::Uninitialized)
        } else {
            // Magic is not MAGIC and not zero
            Err(QueError::CorruptionDetected)
        }
    }

    /// Attempts to read the next element. Returns `None` if the
    /// consumer is caught up.
    pub fn pop(&mut self) -> Option<T> {
        let head_index = self.head & Self::MODULO_MASK;
        let tail = unsafe {
            (*self.spsc.as_ptr())
                .tail
                .load(Ordering::Acquire)
        };
        let previously_read_or_uninitialized = tail <= self.head;
        // Nothing else to read
        if previously_read_or_uninitialized {
            return None;
        }
        let value = unsafe {
            core::ptr::read(
                (*self.spsc.as_ptr())
                    .buffer
                    .as_ptr()
                    .add(head_index),
            )
        };

        self.head += 1;
        self.items_since_last_sync += 1;
        self.maybe_sync();
        return Some(value);
    }

    /// Attempts to read the next element. Returns `None` if the
    /// consumer is caught up.
    pub fn pop_zerocopy<'a>(
        &'a mut self,
    ) -> Option<Element<'a, M, T, N>> {
        let head_index = self.head & Self::MODULO_MASK;
        let tail = unsafe {
            (*self.spsc.as_ptr())
                .tail
                .load(Ordering::Acquire)
        };
        let previously_read_or_uninitialized = tail <= self.head;
        // Nothing else to read
        if previously_read_or_uninitialized {
            return None;
        }
        let value: &mut T = unsafe {
            &mut *(*self.spsc.as_ptr())
                .buffer
                .as_mut_ptr()
                .add(head_index)
        };
        let element = Element {
            value,
            consumer: self,
        };

        return Some(element);
    }

    /// Increments the consumer heartbeat.
    ///
    /// Can be read by the producer to see that the consumer is still
    /// online if done periodically. Can also be used to ack individual
    /// messages or alert that we've joined.
    pub fn beat(&self) {
        unsafe {
            (*self.spsc.as_ptr())
                .consumer_heartbeat
                .fetch_add(1, Ordering::Release);
        }
    }

    /// Checks if the producer has incremented its heartbeat since last
    /// called. Can be used by the consumer to see if the producer is
    /// still online if done periodically. Can also be used to ack
    /// individual messages or alert that we've joined.
    pub fn producer_heartbeat(&mut self) -> bool {
        let heartbeat = unsafe {
            (*self.spsc.as_ptr())
                .producer_heartbeat
                .load(Ordering::Acquire)
        };

        if heartbeat != self.last_producer_heartbeat {
            self.last_producer_heartbeat = heartbeat;
            true
        } else {
            false
        }
    }

    /// Returns pointer to inner padding.
    ///
    /// User is responsible for safe usage.
    ///
    /// Can be used to store metadata (e.g. hash seed).
    ///
    /// Byte array is 128 byte aligned.
    pub fn get_padding_ptr(&self) -> NonNull<[u8; 112]> {
        unsafe {
            NonNull::new_unchecked(
                self.spsc.cast::<u8>().as_ptr().add(512),
            )
            .cast()
        }
    }

    #[inline(always)]
    fn maybe_sync(&mut self) {
        let do_sync = self.items_since_last_sync >= burst_amount::<N>();

        if do_sync {
            self.items_since_last_sync = 0;
            unsafe {
                (*self.spsc.as_ptr())
                    .head
                    .store(self.head, Ordering::Release);
            }
        }
    }
}

impl<M: ChannelMode<T>, T, const N: usize> Drop for Consumer<M, T, N> {
    fn drop(&mut self) {
        // LocalMode is backed by arc
        if M::BACKED_BY_ARCC {
            unsafe { drop(Arc::from_raw(self.spsc.as_ptr())) }
        }
    }
}

File: c/consumer.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <assert.h>
#include <stdatomic.h>
#include <stddef.h>

#include "spsc.c"

#define CONSUMER_CONCAT3(a, b, c) a##b##c
#define CONSUMER_EXPAND_THEN_CONCAT3(a, b, c) CONSUMER_CONCAT3(a, b, c)
#define CONSUMER_(x) CONSUMER_EXPAND_THEN_CONCAT3(CHANNEL_NAME, _consumer_, x)

typedef struct CONSUMER_(consumer) {
    QUE_(spsc_t) * spsc;                  /* Pointer to shared SPSC */
    uint64_t head;                        /* Local head */
    uint64_t items_since_last_sync;       /* Items consumed since last sync */
    uint64_t consumer_index;              /* Consumer index (for future SPMC) */
    uint64_t last_producer_heartbeat;     /* Last known producer heartbeat */
} CONSUMER_(consumer_t);

static inline int
CONSUMER_(join)( void *shmem_region, CONSUMER_(consumer_t) *consumer ) {

    /* Ensure alignment */
    assert( ((size_t)shmem_region % ALIGNMENT) == 0 ); 

    QUE_(spsc_t) *spsc = (QUE_(spsc_t) *)shmem_region;

    /* Join existing queue */
    if( spsc->magic == MAGIC ) {
        if( spsc->capacity != CHANNEL_N ) {
            fprintf( stderr, "Incorrect capacity\n" );
            return 1;
        }

        consumer->spsc = spsc;
        
        /* Set local head to current tail (assume empty upon joining) */
        uint64_t new_head = atomic_load_explicit( &(spsc->tail.value), memory_order_acquire );
        consumer->head = new_head;
        
        /* Update shared head */
        atomic_store_explicit( &(spsc->head.value), new_head, memory_order_release );
        
        consumer->items_since_last_sync = 0;
        consumer->consumer_index = 0;
        consumer->last_producer_heartbeat = atomic_load_explicit( &(spsc->producer_heartbeat.value),
                                                                  memory_order_acquire );
        return 0;
    }
    
    /* Uninitialized */
    if( spsc->magic == 0 ) {
        fprintf( stderr, "Uninitialized channel\n" );
        return 1;
    }

    fprintf( stderr, "Corruption detected\n" );
    return 1;
}

/* Helper function to sync head with shared memory */
static inline void
CONSUMER_(maybe_sync)( CONSUMER_(consumer_t) *consumer ) {
    uint64_t do_sync = consumer->items_since_last_sync >= burst_amount(CHANNEL_N);
    
    if( do_sync ) {
        consumer->items_since_last_sync = 0;
        atomic_store_explicit( &(consumer->spsc->head.value), consumer->head, memory_order_release );
    }
}

static inline int
CONSUMER_(pop)( CONSUMER_(consumer_t) *consumer, CHANNEL_T *value ) {
    QUE_(spsc_t) *spsc = consumer->spsc;

    uint64_t tail = atomic_load_explicit( &(spsc->tail.value), memory_order_acquire );
    int previously_read_or_uninitialized = tail <= consumer->head;
    
    /* Nothing else to read */
    if( previously_read_or_uninitialized ) {
        return 1;
    }
    
    /* Calculate the index in the circular buffer */
    uint64_t index = consumer->head & (CHANNEL_N - 1);
    
    /* Get buffer pointer (starts at offset 640) */
    void *buffer_base = spsc + 1;
    void *addr = (char *)buffer_base + (index * sizeof(CHANNEL_T));
    
    /* Read value */
    memcpy( value, addr, sizeof(CHANNEL_T) );
    
    /* Success - increment head and sync if needed */
    consumer->head++;
    consumer->items_since_last_sync++;
    CONSUMER_(maybe_sync)(consumer);
    
    return 0;
}

static inline void
CONSUMER_(beat)( CONSUMER_(consumer_t) *consumer ) {
    QUE_(spsc_t) *spsc = consumer->spsc;

    /* Atomically increment the consumer's heartbeat */
    atomic_fetch_add_explicit(&(spsc->consumer_heartbeat.value), 1, memory_order_release);
}

static inline int
CONSUMER_(producer_heartbeat)( CONSUMER_(consumer_t) *consumer ) {
    QUE_(spsc_t) *spsc = consumer->spsc;

    /* Load the current heartbeat of the producer */
    uint64_t heartbeat = atomic_load_explicit( &(spsc->producer_heartbeat.value), memory_order_acquire );

    /* Check if the producer's heartbeat has changed */
    if( heartbeat != consumer->last_producer_heartbeat ) {
        /* Update the last known producer heartbeat */
        consumer->last_producer_heartbeat = heartbeat;
        return 1; /* Producer is active */
    }

    return 0; /* Producer has not updated its heartbeat */
}

/* Get pointer to padding area (can be used for metadata) */
static inline void*
CONSUMER_(get_padding_ptr)( CONSUMER_(consumer_t) *consumer ) {
    return consumer->spsc->padding;
}
